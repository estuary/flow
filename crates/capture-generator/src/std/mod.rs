use async_trait::async_trait;
use http::header::AUTHORIZATION;
use reqwest::{Request, Url};
use schemars::{schema::RootSchema, JsonSchema};
use serde::Deserialize;

use crate::interface::*;

#[derive(Default)]
pub struct BearerTokenAuth {
    config: Option<BearerTokenConfig>,
}
#[derive(Deserialize, JsonSchema)]
pub struct BearerTokenConfig {
    token: String,
}
#[async_trait]
impl Auth for BearerTokenAuth {
    type Config = BearerTokenConfig;

    fn prepare(&mut self, config: BearerTokenConfig) -> Result<(), anyhow::Error> {
        self.config = Some(config);
        Ok(())
    }

    async fn authenticate(&mut self, req: &mut Request) -> Result<(), anyhow::Error> {
        let token = &self.config.as_ref().unwrap().token;
        let headers = req.headers_mut();
        headers.insert(AUTHORIZATION, format!("Bearer {token}").parse().unwrap());
        Ok(())
    }
}

/// A very basic stream that sends a HTTP request to a specific URL.
/// Can be used alongside different paginations to create a stream.
pub struct BasicStream {
    pub key: String,
    pub description: String,
    pub endpoint: String,
    pub method: http::Method,
    pub spec: RootSchema,
}
impl Stream for BasicStream {
    type Config = ();

    fn key(&mut self) -> String {
        self.key.clone()
    }

    fn description(&mut self) -> String {
        self.description.clone()
    }

    fn create_request(&mut self) -> Request {
        let mut request = Request::new(self.method.clone(), Url::parse(&self.endpoint).unwrap());
        request
            .headers_mut()
            .insert("User-Agent", "flow-capture".parse().unwrap());

        let body = request.body_mut();
        *body = Some(vec![].into());

        request
    }

    fn spec(&mut self) -> RootSchema {
        self.spec.clone()
    }
}

pub struct BasicPagination {
    pub page_query_field: String,
    pub current_page_field: String,
    pub max_pages_field: String,
}

#[async_trait]
impl Pagination for BasicPagination {
    type Config = ();

    async fn paginate(
        &mut self,
        req: &mut reqwest::Request,
        last_response: reqwest::Response,
    ) -> Result<bool, anyhow::Error> {
        let last_body = last_response.text().await?;
        println!("last_body {:#?}", last_body);
        let last_body = serde_json::from_str::<serde_json::Value>(&last_body)?;

        let res_current_page = last_body.get(&self.current_page_field).unwrap();
        let res_max_page = last_body.get(&self.max_pages_field).unwrap();

        if res_current_page == res_max_page {
            Ok(false)
        } else {
            let current_page = res_current_page.as_u64().unwrap();
            let next_page = (current_page + 1).to_string();

            req.url_mut()
                .query_pairs_mut()
                .append_pair(&self.page_query_field, &next_page);

            println!("{:#?}", req.url().query());

            Ok(true)
        }
    }
}
