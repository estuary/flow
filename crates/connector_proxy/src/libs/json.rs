use doc::ptr::{Pointer, Token};
use schemars::{schema::RootSchema, JsonSchema};
use serde_json::Value;

// Create the RootSchema given datatype T.
pub fn create_root_schema<T: JsonSchema>() -> RootSchema {
    let mut settings = schemars::gen::SchemaSettings::draft07();
    settings.inline_subschemas = true;
    let generator = schemars::gen::SchemaGenerator::new(settings);
    return generator.into_root_schema_for::<T>();
}

// Extract the sub object keyed at `key` from v if v is an object, and returns both the remainder and the removed value.
pub fn remove_subobject(mut v: Value, key: &str) -> (Option<Value>, Value) {
    let mut sub_object = None;

    if let Value::Object(ref mut m) = v {
        sub_object = m.remove(key)
    }

    (sub_object, v)
}

pub fn tokenize_jsonpointer(ptr: &str) -> Vec<String> {
    Pointer::from_str(&ptr)
        .iter()
        .map(|t| match t {
            // Keep the index and next index for now. Could adjust based on usecases.
            Token::Index(ind) => ind.to_string(),
            Token::Property(prop) => prop.to_string(),
            Token::NextIndex => "-".to_string(),
        })
        .collect()
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_tokenize_jsonpointer() {
        let expected: Vec<String> = vec!["p1", "p2", "56", "p3", "-"]
            .iter()
            .map(|s| s.to_string())
            .collect();

        assert!(expected == tokenize_jsonpointer("/p1/p2/56/p3/-"));
    }
}
