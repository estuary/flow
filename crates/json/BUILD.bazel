load("@rules_rust//rust:defs.bzl", "rust_library", "rust_test", "rust_binary")

package(default_visibility = ["//visibility:public"])

# Export testdata for use by other crates (e.g., doc crate tests)
exports_files(glob(["benches/testdata/**"]))

rust_library(
    name = "json",
    srcs = glob([
        "src/**/*.rs",
    ]),
    crate_root = "src/lib.rs",
    edition = "2024",
    deps = [
        "@crates_io//:addr",
        "@crates_io//:bigdecimal",
        "@crates_io//:bitvec",
        "@crates_io//:iri-string",
        "@crates_io//:itertools",
        "@crates_io//:lazy_static",
        "@crates_io//:percent-encoding",
        "@crates_io//:regex",
        "@crates_io//:serde",
        "@crates_io//:serde_json",
        "@crates_io//:thiserror",
        "@crates_io//:time",
        "@crates_io//:url",
        "@crates_io//:uuid",
    ],
)

rust_test(
    name = "json_lib_test",
    crate = ":json",
    size = "small",
    edition = "2024",
    compile_data = glob(["tests/**/*"]),
    data = glob(["src/**/snapshots/*.snap"]),
    env = {
        "INSTA_WORKSPACE_ROOT": ".",
    },
    deps = [
        "@crates_io//:insta",
        "@crates_io//:glob",
    ],
)

_SCHEMA_TEST_DEPS = [
    ":json",
    "@crates_io//:glob",
    "@crates_io//:itertools",
    "@crates_io//:serde_json",
    "@crates_io//:url",
]

[
    rust_test(
        name = test_name,
        srcs = [
            "tests/{}.rs".format(test_name),
            "tests/utils.rs",
        ],
        size = "small",
        data = glob(["tests/**/*.json"]),
        edition = "2024",
        deps = _SCHEMA_TEST_DEPS,
    )
    for test_name in [
        "draft2019_tests",
        "draft2019_format_tests",
        "draft2020_tests",
        "draft2020_format_tests",
    ]
]

_BENCH_DEPS = [
    ":json",
    "@crates_io//:criterion",
    "@crates_io//:serde_json",
    "@crates_io//:url",
]

[
    rust_binary(
        name = "{}_bench".format(bench_name),
        srcs = ["benches/{}.rs".format(bench_name)],
        compile_data = glob(["benches/testdata/**"]),
        edition = "2024",
        deps = _BENCH_DEPS,
    )
    for bench_name in [
        "github_events",
        "citi_rides",
    ]
]
