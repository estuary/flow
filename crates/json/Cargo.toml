[package]
name = "json"
version.workspace = true
rust-version.workspace = true
edition.workspace = true
authors.workspace = true
homepage.workspace = true
repository.workspace = true
license.workspace = true

[dependencies]
addr = { workspace = true }
bigdecimal = { workspace = true }
bitvec = { workspace = true }
fancy-regex = { workspace = true }
fxhash = { workspace = true }
iri-string = { workspace = true }
itertools = { workspace = true }
lazy_static = { workspace = true }
percent-encoding = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
thiserror = { workspace = true }
time = { workspace = true }
tracing = { workspace = true }
url = { workspace = true }
uuid = { workspace = true }

[dev-dependencies]
criterion = { workspace = true }
glob = { workspace = true }

[[bench]]
name = "github_events"
harness = false

[[bench]]
name = "citi_rides"
harness = false
