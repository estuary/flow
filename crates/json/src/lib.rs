pub mod location;
pub mod node;
mod number;
pub mod ptr;
pub mod schema;
pub mod scope;
pub mod validator;

pub use location::{LocatedItem, LocatedProperty, Location};
pub use node::{AsNode, Field, Fields, Node};
pub use number::Number;
pub use ptr::Pointer;
pub use schema::Schema;
pub use scope::Scope;
pub use validator::Validator;
