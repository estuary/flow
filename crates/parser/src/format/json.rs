//! Parser for the json format. This will accept any stream of JSON values separated by whitespace.
//! It allows any amount of whitespace (including newlines) within and in between records.
use super::{Input, Output, ParseError, Parser};
use crate::ParseConfig;

struct JsonParser;

pub fn new_parser() -> Box<dyn Parser> {
    Box::new(JsonParser)
}

impl Parser for JsonParser {
    fn parse(&self, _config: &ParseConfig, content: Input) -> Result<Output, ParseError> {
        // The JSON RFC (RFC 4627) specifies that JSON content is "unicode", but explicitly allows
        // for UTF-16 and 32 encoding schemes in addition to the default of UTF-8. Technically,
        // we're being too permissive here since `transcode_non_utf8` will accept basically any
        // encoding, but I don't see the value of restricting the possible source encodings.
        // Unfortunately, encoding_rs doesn't actually handle utf-32, though
        let input = content
            // only look at the first 32 bytes to determine encoding, since it's most likely only
            // going to look for a BOM.
            .transcode_non_utf8(None, 32)?
            .into_buffered_stream(8192);
        let deser = serde_json::de::Deserializer::from_reader(input).into_iter();
        let wrapped = deser.map(|r| r.map_err(Into::into));
        Ok(Box::new(wrapped))
    }
}
