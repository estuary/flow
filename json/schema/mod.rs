use crate::Number;
use serde_json as sj;
use std::fmt::Write;

pub mod build;
pub mod index;
pub mod intern;
pub mod keywords;
pub mod types;

pub use build::Error as BuildError;

#[derive(Debug)]
pub struct Schema<A>
where
    A: Annotation,
{
    // Canonical URI of this Schema.
    pub curi: url::Url,
    // Keywords of the Schema.
    pub kw: Vec<Keyword<A>>,
    // Interned property names of this Schema.
    pub tbl: intern::Table,
}

/// Annotation is a parsed JSON-Schema annotation that's associated with a Schema instance.
/// An Annotation may wrap, and is potentially convertible to a CoreAnnotation.
pub trait Annotation: Sized + std::fmt::Debug {
    fn as_core(&self) -> Option<&CoreAnnotation>;
}

/// CoreAnnotation represents annotations of the JSON-Schema validation specification.
/// C.f. https://json-schema.org/draft/2019-09/json-schema-validation.html#rfc.section.9
#[derive(Debug)]
pub enum CoreAnnotation {
    Title(String),
    Description(String),
    Default(sj::Value),
    Deprecated(bool),
    ReadOnly(bool),
    WriteOnly(bool),
    Examples(Vec<sj::Value>),
    ContentEncodingBase64,
    ContentMediaType(String),
    /// The "format" annotation can be simply informational, or it can optionally be an assertion
    /// that a string is valid for the given format. We currently only collect and use it as an
    /// annotation, but we may also use it as an assertion in the future.
    /// https://json-schema.org/draft/2019-09/json-schema-validation.html#format
    Format(String),
}
// CoreAnnotation trivially implements Annotation.
impl Annotation for CoreAnnotation {
    fn as_core(&self) -> Option<&CoreAnnotation> {
        Some(self)
    }
}

#[derive(Debug)]
pub enum Keyword<A>
where
    A: Annotation,
{
    // recursiveAnchor indicates that, should this schema appears first in
    // the current *dynamic* scope, then its base URI should be used when
    // resolving a $recursiveRef of a sub-schema of the current scope.
    RecursiveAnchor,
    // $anchor keyword indicates that this Schema should be indexed under
    // an additional canonical URI, which is computed as the base URI
    // extended with a URI fragment composed of the Anchor string.
    Anchor(url::Url),
    // Application of an in-place or child Schema, with respect to this Schema.
    Application(Application, Schema<A>),
    // Validation keyword verified by this Schema.
    Validation(Validation),
    // Annotation collected by a successful application of this Schema.
    Annotation(A),
}

#[derive(Debug)]
pub enum Application {
    // $def is a keyword which defines a schema playing no direct
    // role in validation, but which may be referenced by other schemas
    // (and is indexed).
    Def {
        key: String,
    },

    // In-place applications.
    Ref(url::Url),
    RecursiveRef(String),
    AnyOf {
        index: usize,
    },
    AllOf {
        index: usize,
    },
    OneOf {
        index: usize,
    },
    Not,
    If,
    Then,
    Else,
    DependentSchema {
        if_: String,
        if_interned: intern::Set,
    },

    // Property applications.
    PropertyNames,
    Properties {
        name: String,
        name_interned: intern::Set,
    },
    PatternProperties {
        re: regex::Regex,
    },
    AdditionalProperties,
    UnevaluatedProperties,

    // Item applications.
    Contains,
    Items {
        index: Option<usize>,
    },
    AdditionalItems,
    UnevaluatedItems,
}

impl Application {
    /// Returns a new Location that extends this one with the Application's keyword.
    pub fn push_keyword<'a>(&'a self, parent: &'a super::Location<'a>) -> super::Location<'a> {
        use Application::*;
        match self {
            Def { .. } => parent.push_prop(keywords::DEF),

            // In-place keywords.
            Ref(_) => parent.push_prop(keywords::REF),
            RecursiveRef(_) => parent.push_prop(keywords::RECURSIVE_REF),
            AnyOf { .. } => parent.push_prop(keywords::ANY_OF),
            AllOf { .. } => parent.push_prop(keywords::ALL_OF),
            OneOf { .. } => parent.push_prop(keywords::ONE_OF),
            Not => parent.push_prop(keywords::NOT),
            If => parent.push_prop(keywords::IF),
            Then => parent.push_prop(keywords::THEN),
            Else => parent.push_prop(keywords::ELSE),
            DependentSchema { .. } => parent.push_prop(keywords::DEPENDENT_SCHEMAS),

            // Property keywords.
            PropertyNames => parent.push_prop(keywords::PROPERTY_NAMES),
            Properties { .. } => parent.push_prop(keywords::PROPERTIES),
            PatternProperties { .. } => parent.push_prop(keywords::PATTERN_PROPERTIES),
            AdditionalProperties => parent.push_prop(keywords::ADDITIONAL_PROPERTIES),
            UnevaluatedProperties => parent.push_prop(keywords::UNEVALUATED_PROPERTIES),

            // Item keywords.
            Contains => parent.push_prop(keywords::CONTAINS),
            Items { .. } => parent.push_prop(keywords::ITEMS),
            AdditionalItems => parent.push_prop(keywords::ADDITIONAL_ITEMS),
            UnevaluatedItems => parent.push_prop(keywords::UNEVALUATED_ITEMS),
        }
    }

    /// Returns a new Location that extends this one with the Application's target,
    /// if applicable. If not applicable, a copy of |parent| is returned instead.
    /// The parent should be a Location of this Application's keyword (c.f. push_keyword).
    pub fn push_keyword_target<'a>(
        &'a self,
        parent: &'a super::Location<'a>,
    ) -> super::Location<'a> {
        use Application::*;
        match self {
            Def { key } => parent.push_prop(key),

            // In-place keywords.
            Ref(_) => *parent,
            RecursiveRef(_) => *parent,
            AnyOf { index } => parent.push_item(*index),
            AllOf { index } => parent.push_item(*index),
            OneOf { index } => parent.push_item(*index),
            Not | If | Then | Else => *parent,
            DependentSchema { if_, .. } => parent.push_prop(if_),

            // Property keywords.
            PropertyNames => *parent,
            Properties { name, .. } => parent.push_prop(name),
            PatternProperties { re, .. } => parent.push_prop(re.as_str()),
            AdditionalProperties | UnevaluatedProperties => *parent,

            // Item keywords.
            Contains => *parent,
            Items { index: None } => *parent,
            Items { index: Some(i) } => parent.push_item(*i),
            AdditionalItems | UnevaluatedItems => *parent,
        }
    }

    /// Extend |ptr| with the JSON-Pointer components of this Application.
    pub fn extend_fragment_pointer(&self, mut ptr: String) -> String {
        let l_root = super::Location::Root;
        let l_kw = self.push_keyword(&l_root);
        let l_kwt = self.push_keyword_target(&l_kw);

        write!(&mut ptr, "{}", l_kwt.url_escaped()).unwrap();
        ptr
    }
}

#[derive(Debug)]
pub struct HashedLiteral {
    pub hash: u64,
    pub value: sj::Value,
}

#[derive(Debug)]
pub enum Validation {
    False,
    Type(types::Set),
    Const(HashedLiteral),
    Enum {
        variants: Vec<HashedLiteral>,
    },

    // String-specific validations.
    MaxLength(usize),
    MinLength(usize),
    Pattern(regex::Regex),
    // Format(String),

    // Number-specific validations.
    MultipleOf(Number),
    Maximum(Number),
    ExclusiveMaximum(Number),
    Minimum(Number),
    ExclusiveMinimum(Number),

    // Array-specific validations.
    MaxItems(usize),
    MinItems(usize),
    UniqueItems,
    MaxContains(usize),
    MinContains(usize),

    // Object-specific validations.
    MaxProperties(usize),
    MinProperties(usize),
    Required {
        props: Vec<String>,
        props_interned: intern::Set,
    },
    DependentRequired {
        if_: String,
        if_interned: intern::Set,
        then_: Vec<String>,
        then_interned: intern::Set,
    },
}
