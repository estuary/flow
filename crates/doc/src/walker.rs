use super::{AsNode, Field, Fields, Node};

use fxhash::{hash64, FxHasher64};
use json::Span;
use std::hash::{Hash, Hasher};

/// walk_document is an integration joint between the (newer) AsNode trait
/// and it's representations, and the (older) Walker trait used to walk documents
/// in the course of schema validation.
///
/// Eventually we may want to update our JSON schema validation implementation
/// to directly use the AsNode trait, removing the need for this function.
pub fn walk_document<'l, N: AsNode, W: json::Walker>(
    document: &N,
    walker: &mut W,
    location: &json::Location<'l>,
    span_begin: usize,
) -> json::Span {
    match document.as_node() {
        Node::Array(arr) => {
            let mut span = Span::new(span_begin, ARRAY_SEED);
            let mut hasher = FxHasher64::default();

            for (index, item) in arr.iter().enumerate() {
                let item_span = Span::new(span.end, 0);
                let item_loc = json::LocatedItem {
                    parent: location,
                    index,
                };
                walker.push_item(&item_span, &item_loc);

                let item_loc = json::Location::Item(item_loc);
                let sub_span = walk_document(item, walker, &item_loc, span.end);

                hasher.write_u64(sub_span.hashed);
                span.end = sub_span.end;
            }

            // Hash of this span is the result of hashing over the ordered
            // hashes of each sub-span.
            span.hashed ^= hasher.finish();

            walker.pop_array(&span, location, arr.len());
            span
        }
        Node::Bool(v) => {
            let span = Span::new(span_begin, if v { BOOL_TRUE_HASH } else { BOOL_FALSE_HASH });
            walker.pop_bool(&span, location, v);
            span
        }
        Node::Bytes(_b) => {
            unimplemented!("bytes are not supported for validation yet")
        }
        Node::Null => {
            let span = Span::new(span_begin, UNIT_HASH);
            walker.pop_null(&span, location);
            span
        }
        Node::Float(f) => {
            let hash = {
                // Separately hash integral and fractional hash parts to maintain equality
                // between integer f64 values and u64/i64 types.
                let vt = f.trunc();
                hash64(&(vt as i64)) ^ hash64(&(f - vt).to_bits())
            };
            let span = Span::new(span_begin, hash);
            walker.pop_numeric(&span, location, json::Number::Float(f));
            span
        }
        Node::NegInt(s) => {
            let span = Span::new(span_begin, hash64(&s));
            walker.pop_numeric(&span, location, json::Number::Signed(s));
            span
        }
        Node::PosInt(u) => {
            let span = Span::new(span_begin, hash64(&u));
            walker.pop_numeric(&span, location, json::Number::Unsigned(u));
            span
        }
        Node::String(v) => {
            let span = Span::new(span_begin, STRING_SEED ^ hash64(v));
            walker.pop_str(&span, location, v);
            span
        }
        Node::Object(fields) => {
            let mut span = Span::new(span_begin, OBJECT_SEED);

            for (index, field) in fields.iter().enumerate() {
                let prop_loc = json::LocatedProperty {
                    parent: location,
                    name: field.property(),
                    index,
                };
                let prop_span = Span::new(span.end, hash64(prop_loc.name));
                walker.push_property(&prop_span, &prop_loc);

                let prop_loc = json::Location::Property(prop_loc);
                let sub_span = walk_document(field.value(), walker, &prop_loc, span.end);

                span.end = sub_span.end;

                // Update the hash of our span by XOR'ing in a composed hash
                // of the property name and sub-span value. The XOR is required
                // in order to produce hash values which are invariant to the
                // order in which properties are enumerated.
                let mut h = FxHasher64::default();
                prop_span.hashed.hash(&mut h);
                sub_span.hashed.hash(&mut h);
                span.hashed ^= h.finish();
            }

            walker.pop_object(&span, location, fields.len());
            span
        }
    }
}

// Seeds to distinguish zero-valued types from one another.
// Numbers are not seeded.
const UNIT_HASH: u64 = 0xe0fe5d21a7c19aeb;
const BOOL_TRUE_HASH: u64 = 0x3bd83018139b2c4d;
const BOOL_FALSE_HASH: u64 = 0x4cd6d6c279e081c0;
const ARRAY_SEED: u64 = 0xca910bdc0b6441dd;
const OBJECT_SEED: u64 = 0x76662a22f0a45102;
const STRING_SEED: u64 = 0x5570bb24d6cdeee2;

#[cfg(test)]
mod test {

    use super::walk_document;
    use crate::HeapNode;
    use json::{LocatedItem, LocatedProperty, Location, Number, Span};

    struct MockWalker {
        calls: Vec<String>,
    }

    impl json::Walker for MockWalker {
        fn push_property<'a>(&mut self, span: &Span, loc: &'a LocatedProperty<'a>) {
            self.calls.push(format!(
                "push_property(span: {span:?}, loc: {})",
                json::Location::Property(*loc).pointer_str().to_string()
            ))
        }
        fn push_item<'a>(&mut self, span: &Span, loc: &'a LocatedItem<'a>) {
            self.calls.push(format!(
                "push_item(span: {span:?}, loc: {})",
                json::Location::Item(*loc).pointer_str().to_string()
            ))
        }
        fn pop_object<'a>(&mut self, span: &Span, loc: &'a Location<'a>, num_properties: usize) {
            self.calls.push(format!(
                "pop_object(span: {span:?}, loc: {}, num_properties: {num_properties})",
                loc.pointer_str().to_string()
            ))
        }
        fn pop_array<'a>(&mut self, span: &Span, loc: &'a Location<'a>, num_items: usize) {
            self.calls.push(format!(
                "pop_array(span: {span:?}, loc: {}, num_items: {num_items})",
                loc.pointer_str().to_string()
            ))
        }
        fn pop_bool<'a>(&mut self, span: &Span, loc: &'a Location<'a>, val: bool) {
            self.calls.push(format!(
                "pop_bool(span: {span:?}, loc: {}, val: {val})",
                loc.pointer_str().to_string()
            ))
        }
        fn pop_numeric<'a>(&mut self, span: &Span, loc: &'a Location<'a>, val: Number) {
            self.calls.push(format!(
                "pop_numeric(span: {span:?}, loc: {}, val: {val})",
                loc.pointer_str().to_string()
            ))
        }
        fn pop_str<'a>(&mut self, span: &Span, loc: &'a Location<'a>, val: &'a str) {
            self.calls.push(format!(
                "pop_str(span: {span:?}, loc: {}, val: {val})",
                loc.pointer_str().to_string()
            ))
        }
        fn pop_null<'a>(&mut self, span: &Span, loc: &'a Location<'a>) {
            self.calls.push(format!(
                "pop_null(span: {span:?}, loc: {})",
                loc.pointer_str().to_string()
            ))
        }
    }

    #[test]
    fn test_walk_regression() {
        let fixture = serde_json::json!({
            "numbers": [ 0x1111111111111111 as u64, -1234, 56.7891122334455],
            "repeat": "repeat",
            "some": {"bytes":"c29tZSBieXRlcw=="},
            "null": null,
            "foo": {
                "true": true,
                "two": 2,
                "repeat": {"repeat": "repeat"},
            },
            "bar": {
                "false": false,
                "flat-array": [{"object":"value", "ninish": 9.999}, 32, {"third": 3}],
                "nested-array": [[[1,2],[3],4],5,[6]]
            }
        });

        let alloc = HeapNode::new_allocator();
        let doc = HeapNode::from_serde(&fixture, &alloc).unwrap();

        let mut walker = MockWalker { calls: Vec::new() };

        // Uncomment me to run the fixture through the original deserialization-based
        // Walker implementation. This ensures that that implementation and this one
        // produce identical outputs.
        //json::de::walk(&fixture, &mut walker).unwrap();

        walk_document(&doc, &mut walker, &json::Location::Root, 0);

        insta::assert_debug_snapshot!(&walker.calls);
    }
}
