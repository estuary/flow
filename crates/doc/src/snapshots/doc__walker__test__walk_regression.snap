---
source: crates/doc/src/walker.rs
expression: "&walker.calls"
---
[
    "push_property(span: Span { begin: 1, end: 2, hashed: 3492387289669676943 }, loc: /bar)",
    "push_property(span: Span { begin: 2, end: 3, hashed: 9111827012676064624 }, loc: /bar/false)",
    "pop_bool(span: Span { begin: 2, end: 3, hashed: 5536848922655359424 }, loc: /bar/false, val: false)",
    "push_property(span: Span { begin: 3, end: 4, hashed: 12951238764204348148 }, loc: /bar/flat-array)",
    "push_item(span: Span { begin: 4, end: 5, hashed: 0 }, loc: /bar/flat-array/0)",
    "push_property(span: Span { begin: 5, end: 6, hashed: 12359339792617410391 }, loc: /bar/flat-array/0/ninish)",
    "pop_numeric(span: Span { begin: 5, end: 6, hashed: 308683169865322957 }, loc: /bar/flat-array/0/ninish, val: 9.999)",
    "push_property(span: Span { begin: 6, end: 7, hashed: 17066697293871039574 }, loc: /bar/flat-array/0/object)",
    "pop_str(span: Span { begin: 6, end: 7, hashed: 4779665107170170139 }, loc: /bar/flat-array/0/object, val: value)",
    "pop_object(span: Span { begin: 4, end: 7, hashed: 6712404287633763183 }, loc: /bar/flat-array/0, num_properties: 2)",
    "push_item(span: Span { begin: 7, end: 8, hashed: 0 }, loc: /bar/flat-array/1)",
    "pop_numeric(span: Span { begin: 7, end: 8, hashed: 3429551472952562336 }, loc: /bar/flat-array/1, val: 32)",
    "push_item(span: Span { begin: 8, end: 9, hashed: 0 }, loc: /bar/flat-array/2)",
    "push_property(span: Span { begin: 9, end: 10, hashed: 686794948548171365 }, loc: /bar/flat-array/2/third)",
    "pop_numeric(span: Span { begin: 9, end: 10, hashed: 17615343019692007359 }, loc: /bar/flat-array/2/third, val: 3)",
    "pop_object(span: Span { begin: 8, end: 10, hashed: 3347361485416429113 }, loc: /bar/flat-array/2, num_properties: 1)",
    "pop_array(span: Span { begin: 3, end: 10, hashed: 7969316532600871641 }, loc: /bar/flat-array, num_items: 3)",
    "push_property(span: Span { begin: 10, end: 11, hashed: 6050281313959441828 }, loc: /bar/nested-array)",
    "push_item(span: Span { begin: 11, end: 12, hashed: 0 }, loc: /bar/nested-array/0)",
    "push_item(span: Span { begin: 12, end: 13, hashed: 0 }, loc: /bar/nested-array/0/0)",
    "push_item(span: Span { begin: 13, end: 14, hashed: 0 }, loc: /bar/nested-array/0/0/0)",
    "pop_numeric(span: Span { begin: 13, end: 14, hashed: 5871781006564002453 }, loc: /bar/nested-array/0/0/0, val: 1)",
    "push_item(span: Span { begin: 14, end: 15, hashed: 0 }, loc: /bar/nested-array/0/0/1)",
    "pop_numeric(span: Span { begin: 14, end: 15, hashed: 11743562013128004906 }, loc: /bar/nested-array/0/0/1, val: 2)",
    "pop_array(span: Span { begin: 12, end: 15, hashed: 3395477536205329037 }, loc: /bar/nested-array/0/0, num_items: 2)",
    "push_item(span: Span { begin: 15, end: 16, hashed: 0 }, loc: /bar/nested-array/0/1)",
    "push_item(span: Span { begin: 16, end: 17, hashed: 0 }, loc: /bar/nested-array/0/1/0)",
    "pop_numeric(span: Span { begin: 16, end: 17, hashed: 17615343019692007359 }, loc: /bar/nested-array/0/1/0, val: 3)",
    "pop_array(span: Span { begin: 15, end: 17, hashed: 5640907580079321590 }, loc: /bar/nested-array/0/1, num_items: 1)",
    "push_item(span: Span { begin: 17, end: 18, hashed: 0 }, loc: /bar/nested-array/0/2)",
    "pop_numeric(span: Span { begin: 17, end: 18, hashed: 5040379952546458196 }, loc: /bar/nested-array/0/2, val: 4)",
    "pop_array(span: Span { begin: 11, end: 18, hashed: 1245630777918685515 }, loc: /bar/nested-array/0, num_items: 3)",
    "push_item(span: Span { begin: 18, end: 19, hashed: 0 }, loc: /bar/nested-array/1)",
    "pop_numeric(span: Span { begin: 18, end: 19, hashed: 10912160959110460649 }, loc: /bar/nested-array/1, val: 5)",
    "push_item(span: Span { begin: 19, end: 20, hashed: 0 }, loc: /bar/nested-array/2)",
    "push_item(span: Span { begin: 20, end: 21, hashed: 0 }, loc: /bar/nested-array/2/0)",
    "pop_numeric(span: Span { begin: 20, end: 21, hashed: 16783941965674463102 }, loc: /bar/nested-array/2/0, val: 6)",
    "pop_array(span: Span { begin: 19, end: 21, hashed: 14060822398349058443 }, loc: /bar/nested-array/2, num_items: 1)",
    "pop_array(span: Span { begin: 10, end: 21, hashed: 6169813963366165170 }, loc: /bar/nested-array, num_items: 3)",
    "pop_object(span: Span { begin: 1, end: 21, hashed: 8400310503551780749 }, loc: /bar, num_properties: 3)",
    "push_property(span: Span { begin: 21, end: 22, hashed: 17561646994490527473 }, loc: /foo)",
    "push_property(span: Span { begin: 22, end: 23, hashed: 8515467036952247710 }, loc: /foo/repeat)",
    "push_property(span: Span { begin: 23, end: 24, hashed: 8515467036952247710 }, loc: /foo/repeat/repeat)",
    "pop_str(span: Span { begin: 23, end: 24, hashed: 2548402987741205372 }, loc: /foo/repeat/repeat, val: repeat)",
    "pop_object(span: Span { begin: 22, end: 24, hashed: 879477391208173632 }, loc: /foo/repeat, num_properties: 1)",
    "push_property(span: Span { begin: 24, end: 25, hashed: 13840727042744474435 }, loc: /foo/true)",
    "pop_bool(span: Span { begin: 24, end: 25, hashed: 4312249523173534797 }, loc: /foo/true, val: true)",
    "push_property(span: Span { begin: 25, end: 26, hashed: 15530891139749574631 }, loc: /foo/two)",
    "pop_numeric(span: Span { begin: 25, end: 26, hashed: 11743562013128004906 }, loc: /foo/two, val: 2)",
    "pop_object(span: Span { begin: 21, end: 26, hashed: 12625743532182824938 }, loc: /foo, num_properties: 3)",
    "push_property(span: Span { begin: 26, end: 27, hashed: 6415180377498958507 }, loc: /null)",
    "pop_null(span: Span { begin: 26, end: 27, hashed: 16212498107710151403 }, loc: /null)",
    "push_property(span: Span { begin: 27, end: 28, hashed: 15110081691896058520 }, loc: /numbers)",
    "push_item(span: Span { begin: 28, end: 29, hashed: 0 }, loc: /numbers/0)",
    "pop_numeric(span: Span { begin: 28, end: 29, hashed: 3297896747637643493 }, loc: /numbers/0, val: 1229782938247303441)",
    "push_item(span: Span { begin: 29, end: 30, hashed: 0 }, loc: /numbers/1)",
    "pop_numeric(span: Span { begin: 29, end: 30, hashed: 3792658867874758086 }, loc: /numbers/1, val: -1234)",
    "push_item(span: Span { begin: 30, end: 31, hashed: 0 }, loc: /numbers/2)",
    "pop_numeric(span: Span { begin: 30, end: 31, hashed: 18343232044205362776 }, loc: /numbers/2, val: 56.7891122334455)",
    "pop_array(span: Span { begin: 27, end: 31, hashed: 15174838816573947911 }, loc: /numbers, num_items: 3)",
    "push_property(span: Span { begin: 31, end: 32, hashed: 8515467036952247710 }, loc: /repeat)",
    "pop_str(span: Span { begin: 31, end: 32, hashed: 2548402987741205372 }, loc: /repeat, val: repeat)",
    "push_property(span: Span { begin: 32, end: 33, hashed: 16388633398439343333 }, loc: /some)",
    "push_property(span: Span { begin: 33, end: 34, hashed: 11802959503358919994 }, loc: /some/bytes)",
    "pop_str(span: Span { begin: 33, end: 34, hashed: 436868375673068097 }, loc: /some/bytes, val: c29tZSBieXRlcw==)",
    "pop_object(span: Span { begin: 32, end: 34, hashed: 7824292072841271611 }, loc: /some, num_properties: 1)",
    "pop_object(span: Span { begin: 0, end: 34, hashed: 3550334853817554370 }, loc: , num_properties: 6)",
]
